src/libsync/discovery.cpp

```diff
@@ -0,0 +1,1218 @@
+                auto filePerms = base._remotePerm; // Technicly we should use the one from the server, but we'll assume it is the same
+                //true when it is just a rename in the same directory. (not a move)
+                bool isRename = originalPath.startsWith(_currentFolder._original)
+                    && originalPath.lastIndexOf('/') == _currentFolder._original.size();
+                // Check if we are allowed to move to the destination.
+                bool destinationOK = true;
+                if (isRename || destPerms.isNull()) {
+                    // no need to check for the destination dir permission
+                    destinationOK = true;
+                } else if (item->isDirectory() && !destPerms.hasPermission(RemotePermissions::CanAddSubDirectories)) {
+                    destinationOK = false;
+                } else if (!item->isDirectory() && !destPerms.hasPermission(RemotePermissions::CanAddFile)) {
+                    destinationOK = false;
+                }
+
+                // check if we are allowed to move from the source
+                bool sourceOK = true;
+                if (!filePerms.isNull()
+                    && ((isRename && !filePerms.hasPermission(RemotePermissions::CanRename))
+                           || (!isRename && !filePerms.hasPermission(RemotePermissions::CanMove)))) {
+                    // We are not allowed to move or rename this file
+                    sourceOK = false;
+                }
+                if (!sourceOK || !destinationOK) {
+                    qCInfo(lcDisco) << "Not a move because permission does not allow it." << sourceOK << destinationOK;
+                    if (!sourceOK) {
+                        // This is the behavior that we had in the client <= 2.5.
+                        _discoveryData->_statedb->avoidRenamesOnNextSync(base._path);
+                    }
+                    isMove = false;
+                }
+            }
+
+            if (isMove) {
+                QByteArray oldEtag;
+                auto it = _discoveryData->_deletedItem.find(originalPath);
+                if (it != _discoveryData->_deletedItem.end()) {
+                    if ((*it)->_instruction != CSYNC_INSTRUCTION_REMOVE && (*it)->_type != ItemTypeVirtualFile)
+                        isMove = false;
+                    else
+                        (*it)->_instruction = CSYNC_INSTRUCTION_NONE;
+                    oldEtag = (*it)->_etag;
+                    if (!item->isDirectory() && oldEtag != base._etag) {
+                        isMove = false;
+                    }
+                }
+                if (auto deleteJob = _discoveryData->_queuedDeletedDirectories.value(originalPath)) {
+                    oldEtag = deleteJob->_dirItem->_etag;
+                    delete _discoveryData->_queuedDeletedDirectories.take(originalPath);
+                }
```

Comment:
There's a good chance we haven't visited originalPath yet, and I'm assuming the way it works is that when we get to it but have the _renamedItems entry, we won't set it up for REMOVE. If true, please add a comment to that effect here.